File: LLD/Design-Patterns/Adapter/code/Adapter.java
// Adapter Pattern — adapting an incompatible third-party payment SDK.
// Compile: javac Adapter.java
// Run:     java Adapter

// The interface our application code expects everywhere.
interface PaymentProcessor {
    void pay(double amountDollars);
}

// A "native" implementation that already matches our interface.
class InHousePaymentProcessor implements PaymentProcessor {
    public void pay(double amountDollars) {
        System.out.println("In-house processor charged $" + amountDollars);
    }
}

// The third-party SDK we don't control — incompatible method name and units.
class ThirdPartySDK {
    void makeTransaction(int amountCents, String currencyCode) {
        System.out.println("ThirdPartySDK processed " + amountCents + " " + currencyCode + " cents");
    }
}

// The Adapter: implements OUR interface, translates to the SDK's shape internally.
class ThirdPartyPaymentAdapter implements PaymentProcessor {
    private final ThirdPartySDK sdk;

    ThirdPartyPaymentAdapter(ThirdPartySDK sdk) {
        this.sdk = sdk;
    }

    public void pay(double amountDollars) {
        int cents = (int) Math.round(amountDollars * 100);
        sdk.makeTransaction(cents, "USD");
    }
}

public class Adapter {
    static void checkout(PaymentProcessor processor, double amount) {
        processor.pay(amount);
    }

    public static void main(String[] args) {
        PaymentProcessor inHouse = new InHousePaymentProcessor();
        PaymentProcessor thirdParty = new ThirdPartyPaymentAdapter(new ThirdPartySDK());

        checkout(inHouse, 49.99);
        checkout(thirdParty, 49.99);
    }
}

File: LLD/Design-Patterns/Adapter/code/adapter.js
/**
 * Adapter Pattern — adapting an incompatible third-party payment SDK.
 * Run: node adapter.js
 */

class InHousePaymentProcessor {
  pay(amountDollars) {
    console.log(`In-house processor charged $${amountDollars}`);
  }
}

// The third-party SDK we don't control — incompatible method name and units.
class ThirdPartySDK {
  makeTransaction(amountCents, currencyCode) {
    console.log(`ThirdPartySDK processed ${amountCents} ${currencyCode} cents`);
  }
}

// The Adapter: exposes the same pay() shape our app expects, translates internally.
class ThirdPartyPaymentAdapter {
  constructor(sdk) {
    this.sdk = sdk;
  }

  pay(amountDollars) {
    const cents = Math.round(amountDollars * 100);
    this.sdk.makeTransaction(cents, "USD");
  }
}

function checkout(processor, amount) {
  processor.pay(amount);
}

const inHouse = new InHousePaymentProcessor();
const thirdParty = new ThirdPartyPaymentAdapter(new ThirdPartySDK());

checkout(inHouse, 49.99);
checkout(thirdParty, 49.99);

module.exports = { InHousePaymentProcessor, ThirdPartyPaymentAdapter };
